File: moderate/card_number.rb
File.open(ARGV[0]).each_line do |line|
  s = line.delete(" \n").split('').map(&:to_i)
  (s.length - 2).step(0, -2) do |i|
    s[i] *= 2
    s[i] = s[i] % 10 + 1 if s[i] > 9
  end
  puts s.inject(:+) % 10 == 0 ? 1 : 0
end

File: moderate/card_number.pl
use strict;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    next if ( $line =~ m/^\s$/ );
    chomp($line);
    $line =~ s/ //g;
    my @s = split( //, $line );
    for ( my $i = scalar @s - 2 ; $i >= 0 ; $i -= 2 ) {
        @s[$i] <<= 1;
        @s[$i] = @s[$i] % 10 + 1 if @s[$i] > 9;
    }
    my $r;
    map { $r += $_ } @s;
    print( ( $r % 10 == 0 ) ? "1\n" : "0\n" );
}
close(INFILE);
